File: src/main/java/eu/mshade/enderframe/protocol/temp/TempProtocol.java
package eu.mshade.enderframe.protocol.temp;

import eu.mshade.enderframe.protocol.*;
import eu.mshade.enderframe.protocol.packet.PacketOutDisconnect;
import eu.mshade.enderframe.protocol.temp.listener.PacketHandshakeListener;
import eu.mshade.enderframe.protocol.temp.listener.PacketPingListener;
import eu.mshade.enderframe.protocol.temp.listener.PacketStatusListener;
import eu.mshade.enderframe.protocol.temp.packet.*;
import io.netty.buffer.ByteBuf;
import io.netty.channel.Channel;


public class TempProtocol extends Protocol {

    private static TempProtocol INSTANCE;
    public TempProtocol() {
        INSTANCE = this;

        this.getEventBus().subscribe(PacketInHandshake.class, new PacketHandshakeListener());
        this.getEventBus().subscribe(PacketInStatus.class, new PacketStatusListener());
        this.getEventBus().subscribe(PacketInPing.class, new PacketPingListener());

        this.getProtocolRegistry().registerIn(ProtocolStatus.HANDSHAKE, 0x00, PacketInHandshake.class);

        this.getProtocolRegistry().registerOut(ProtocolStatus.STATUS, 0x00, PacketOutStatus.class);
        this.getProtocolRegistry().registerOut(ProtocolStatus.STATUS, 0x01, PacketOutPong.class);
        this.getProtocolRegistry().registerIn(ProtocolStatus.STATUS, 0x00, PacketInStatus.class);
        this.getProtocolRegistry().registerIn(ProtocolStatus.STATUS, 0x01, PacketInPing.class);

        this.getProtocolRegistry().registerOut(ProtocolStatus.LOGIN, 0x00, PacketOutDisconnect.class);
    }


    @Override
    public ProtocolBuffer getProtocolBuffer(ByteBuf byteBuf) {
        return new TempProtocolBuffer(byteBuf);
    }

    @Override
    public SessionWrapper getSessionWrapper(Channel channel) {
        return new TempSessionWrapper(channel);
    }

    @Override
    public MinecraftProtocolVersion getMinecraftProtocolVersion() {
        return MinecraftProtocolVersion.UNKNOWN;
    }


    public static TempProtocol getInstance(){
        return (INSTANCE != null ? INSTANCE : new TempProtocol());
    }

}

File: src/main/java/eu/mshade/enderframe/protocol/temp/TempProtocol.kt
package eu.mshade.enderframe.protocol.temp

import eu.mshade.enderframe.protocol.*
import eu.mshade.enderframe.protocol.packet.MinecraftPacketOutDisconnect
import eu.mshade.enderframe.protocol.temp.listener.PacketHandshakeListener
import eu.mshade.enderframe.protocol.temp.listener.PacketPingListener
import eu.mshade.enderframe.protocol.temp.listener.PacketStatusListener
import eu.mshade.enderframe.protocol.temp.packet.*
import io.netty.buffer.ByteBuf
import io.netty.channel.Channel

class TempProtocol : Protocol() {

    init {
        getEventBus().subscribe(MinecraftPacketInHandshake::class.java, PacketHandshakeListener())
        getEventBus().subscribe(MinecraftPacketInStatus::class.java, PacketStatusListener())
        getEventBus().subscribe(MinecraftPacketInPing::class.java, PacketPingListener())

        getProtocolRegistry().registerIn(MinecraftProtocolStatus.HANDSHAKE, 0x00, MinecraftPacketInHandshake::class.java)

        getProtocolRegistry().registerOut(MinecraftProtocolStatus.STATUS, 0x00, MinecraftPacketOutStatus::class.java)
        getProtocolRegistry().registerOut(MinecraftProtocolStatus.STATUS, 0x01, MinecraftPacketOutPong::class.java)
        getProtocolRegistry().registerIn(MinecraftProtocolStatus.STATUS, 0x00, MinecraftPacketInStatus::class.java)
        getProtocolRegistry().registerIn(MinecraftProtocolStatus.STATUS, 0x01, MinecraftPacketInPing::class.java)
        getProtocolRegistry().registerOut(MinecraftProtocolStatus.LOGIN, 0x00, MinecraftPacketOutDisconnect::class.java)
    }

    override fun getProtocolBuffer(byteBuf: ByteBuf): ProtocolBuffer {
        return TempProtocolBuffer(byteBuf)
    }

    override fun getSessionWrapper(channel: Channel): MinecraftSession {
        return TempMinecraftSession(channel)
    }

    override fun getMinecraftProtocolVersion(): MinecraftProtocolVersion {
        return MinecraftProtocolVersion.UNKNOWN
    }

    companion object {
        var INSTANCE = TempProtocol()
    }
}